### ISSUE TITLE
Fail to enable YubiKey U2F via OpenSSH: fido_dev_make_cred: FIDO_ERR_RX

### ISSUE THREAD

**Issue by YufeiHu** [Yubico/libfido2#145]
I’m using a macOS Mojave Version 10.14.6. I’ve upgraded my openSSH to the latest version via Homebrew with info shown below:
`OpenSSH_8.2p1, OpenSSL 1.1.1d  10 Sep 2019`

I've also upgraded `libfido2` to version `1.3.1`

My YubiKey model is: YubiKey FIPS
My YubiKey firmware is: 4.4.5

However, when I tried to run:
`ssh-keygen -t ecdsa-sk -vv`

I got a following error message:

```
Generating public/private ecdsa-sk key pair.
You may need to touch your authenticator to authorize key generation.
debug1: start_helper: starting /usr/local/Cellar/openssh/8.2p1_1/libexec/ssh-sk-helper
debug1: sshsk_enroll: provider "internal", device "(null)", application "ssh:", userid "(null)", flags 0x01, challenge len 0
debug1: sshsk_enroll: using random challenge
debug1: ssh_sk_enroll: using device IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS09@14400000/Yubikey 4 OTP+U2F+CCID@14400000/IOUSBHostInterface@1/IOUSBHostHIDDevice@14400000,1
debug1: ssh_sk_enroll: fido_dev_make_cred: FIDO_ERR_RX
debug1: sshsk_enroll: provider "internal" returned failure -1
debug1: ssh-sk-helper: Enrollment failed: invalid format
debug1: ssh-sk-helper: reply len 8
debug1: client_converse: helper returned error -4
Key enrollment failed: invalid format
```

I’ve googled it, there is absolutely zero clues there about `debug1: ssh_sk_enroll: fido_dev_make_cred: FIDO_ERR_RX`. Any clues?

**Comment by martelletto** [comment]
Hi,

Thank you for the detailed report.

What you described looks like https://github.com/Yubico/libfido2/issues/99. There is a fix for it in HEAD (ba9468830). A version of libfido2 with ba9468830 should be out in April.

-p.

**Comment by YufeiHu** [comment]
Hi,

Thanks for the reply.
I've tried compile and install the libfido2 at version ba94688, and provide the `libsk-libfido2.so` file to `ssh-keygen` with the command `ssh-keygen -t ecdsa-sk -w /usr/local/lib/libsk-libfido2.so`

However, it throws me another error:
```
Generating public/private ecdsa-sk key pair.
You may need to touch your authenticator to authorize key generation.
debug1: start_helper: starting /usr/local/Cellar/openssh/8.2p1_1/libexec/ssh-sk-helper
debug1: sshsk_enroll: provider "/usr/local/lib/libsk-libfido2.so", device "(null)", application "ssh:", userid "(null)", flags 0x01, challenge len 0
debug1: sshsk_enroll: using random challenge
debug1: sshsk_open: provider /usr/local/lib/libsk-libfido2.so implements version 0x00020000
Provider "/usr/local/lib/libsk-libfido2.so" implements unsupported version 0x00020000 (supported: 0x00040000)
debug1: ssh-sk-helper: Enrollment failed: invalid format
debug1: ssh-sk-helper: reply len 8
debug1: client_converse: helper returned error -4
Key enrollment failed: invalid format
```

According to [#116](https://github.com/Yubico/libfido2/issues/116), it seems that I have to use the internal middleware in OpenSSH. But their internal middleware will just throw the error message I posted earlier. Do I need to contact OpenSSH developers then? Or did I miss anything?

Thanks

**Comment by martelletto** [comment]
Hi,

OpenSSH's internal middleware is probably picking up Homebrew's libfido2 instead of yours.

It might be easier to use Homebrew's libfido2 and build a patched version of OpenSSH:

- Install pkg-config, autoconf, and automake from Homebrew;
- Make sure pkg-config sees libfido2: pkg-config --list-all | grep libfido2. If it doesn't, you might need to adjust PKG_CONFIG_PATH;
- Clone OpenSSH portable;
- Patch sk-usbhid.c so that there is a call to fido_dev_force_u2f() after every successful call to fido_dev_open():

```diff
diff --git sk-usbhid.c sk-usbhid.c
index ad83054a..04279c4f 100644
--- sk-usbhid.c
+++ sk-usbhid.c
@@ -213,6 +213,7 @@ find_device(const char *path, const uint8_t *message, size_t message_len,
                        fido_dev_free(&dev);
                        return NULL;
                }
+               fido_dev_force_u2f(dev);
                return dev;
        }

@@ -249,6 +250,7 @@ find_device(const char *path, const uint8_t *message, size_t message_len,
                        fido_dev_free(&dev);
                        continue;
                }
+               fido_dev_force_u2f(dev);
                if (try_device(dev, message, message_len, application,
                    key_handle, key_handle_len) == 0) {
                        skdebug(__func__, "found key");
@@ -523,6 +525,7 @@ sk_enroll(uint32_t alg, const uint8_t *challenge, size_t challenge_len,
                skdebug(__func__, "fido_dev_open: %s", fido_strerr(r));
                goto out;
        }
+       fido_dev_force_u2f(dev);
        if ((r = fido_dev_make_cred(dev, cred, pin)) != FIDO_OK) {
                skdebug(__func__, "fido_dev_make_cred: %s", fido_strerr(r));
                ret = fidoerr_to_skerr(r);
```
- Tell SSH where to find libfido2: export CFLAGS=$(pkg-config --cflags libfido2); export LDFLAGS=$(pkg-config --libs libfido2);
- Build SSH: autoreconf -i .; ./configure; make

This should work for the next few weeks, until a new version of libfido2 is out and packaged by Homebrew. Another option - if possible - would be to use a non-FIPS Yubikey; this should work out of the box with what is available in Homebrew.

My sincere apologies for all the confusion.

-p.

**Comment by martelletto** [comment]
Duplicate of #99, for which a fix will be included in libfido2 1.4.0.
